docs/internals/optimizer.rst

```diff
@@ -0,0 +1,1254 @@
+.. index:: optimizer, optimiser, common subexpression elimination, constant propagation
+.. _optimizer:
+
+*************
+The Optimizer
+*************
+
+The Solidity compiler uses two different optimizer modules: The "old" optimizer
+that operates at opcode level and the "new" optimizer that operates on Yul IR code.
+
+The opcode-based optimizer applies simplification rules from the
+`list to opcodes <https://github.com/ethereum/solidity/blob/develop/libevmasm/RuleList.h>`_
+next to each other. It also combines equal code sets and removes unused code.
+
+The Yul-based optimizer is much more powerful, because it can work across function
+calls: In Yul it is not possible to perform arbitrary jumps, so it is for example
+possible to compute the side-effects of each function: If a function does not modify
+storage, a call to it can be swapped with a function that does. If a function is
+side-effect free and its result is multiplied by zero, you can remove the function
+call completely.
+
+Currently, the parameter ``--optimize`` activates the opcode-based optimizer for the
+generated bytecode and the Yul optimizer for the Yul code generated for ABIv2 codec.
+For a Yul-based codegen, one can use ``solc --ir-optimized --optimize`` to produce
+optimized Yul for a Solidity source. Similarly, use ``solc --strict-assembly --optimize``
+for a stand-alone Yul mode.
+
+You can find more details on both optimizer modules and their optimization steps below.
+
+Benefits of Optimizing Solidity Code
+====================================
+
+Overall, the optimizer tries to simplify complicated expressions, which reduces both code
+size and execution cost. It also specializes or inlines functions. Especially
+function inlining is an operation that can cause much bigger code, but it is
+often done because it results in opportunities for more simplifications.
+
+
+Differences between Optimized and Non-Optimized Code
+====================================================
+
+Generally, the most visible difference would be constant expressions getting evaluated.
+When it comes to the ASM output, one can also notice reduction of equivalent/duplicate
+"code blocks" (compare the output of the flags ``--asm`` and ``--asm --optimize``). However,
+when it comes to the Yul/intermediate-representation, there can be significant
+differences, for example, functions can get inlined, combined, rewritten to eliminate
+redundancies, etc. (compare the output between the flags ``--ir`` and
+``--optimize --ir-optimized``).
+
+Optimize Runs
+=============
+
+The number of runs (``--optimize-runs``) specifies roughly how often each opcode of the
+deployed code will be executed across the life-time of the contract. A "runs" parameter
+of "1" will produce short but expensive code. The largest value is ``2**32-1``.
+
```

Comment:
```suggestion
.. note:
  A common misconception is that this parameter specifies the number of iterations of the optimizer.
  This is not true: The optimizer will always run as many times as it can still improve the code.
```